File: Leetcode150/289. Game of Life.js
/**
 * @param {number[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var gameOfLife = function(board) 
{
    const neighbors = [0, 1, -1];
    const rows = board.length;
    const cols = board[0].length;

    //Iterate over the board
    for(let row = 0; row < rows; row++)
    {
        for(let col = 0; col < cols; col++)
        {
            // Count the number of live neighbors
            let liveNeighbors = 0;

            for (let i = 0; i < 3; i++) 
            {
                for (let j = 0; j < 3; j++) 
                {

                    if (!(neighbors[i] === 0 && neighbors[j] === 0)) 
                    {
                        let r = row + neighbors[i];
                        let c = col + neighbors[j];

                        // Check if the neighboring cell is valid and was originally live
                        if (r >= 0 && r < rows && c >= 0 && c < cols && Math.abs(board[r][c]) === 1) 
                            liveNeighbors += 1;

                    }
                }
            }
            // Rule 1 or Rule 3: Any live cell with <2 or >3 live neighbors dies
            if (board[row][col] === 1 && (liveNeighbors < 2 || liveNeighbors > 3)) 
                board[row][col] = -1; // -1 signifies it was originally live but now dead

            // Rule 4: Any dead cell with exactly 3 live neighbors becomes a live cell
            if (board[row][col] === 0 && liveNeighbors === 3) 
                board[row][col] = 2; // 2 signifies it was originally dead but now live
            
        }
    }
    // Convert temporary values (-1 and 2) back to 0 and 1
    for (let row = 0; row < rows; row++) 
    {
        for (let col = 0; col < cols; col++) 
            board[row][col] = board[row][col] > 0 ? 1 : 0;
    }
};

File: Matrix/289. Game of Life.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        rows, cols = len(board), len(board[0])
        directions = [(1, 0), (-1, 0), (1, 1), (-1, -1), (0, 1), (0, -1), (1, -1), (-1, 1)]

        #result = [[0] * cols for _ in range(rows)]
        #result = [[0 for _ in range(cols)] for _ in range(rows)]
        # [0] * n
        # [0 for _ in range(n)]
        # [[0, 1, 0], [0, 1, 0], [0, 1, 0], [0, 1, 0]]

        for row in range(rows):
            for col in range(cols):
                value = board[row][col]
                live_cell = 0
                for (dx, dy) in directions:
                    new_row, new_col = dx + row, dy + col
                    
                    if 0 <= new_row < rows and 0 <= new_col < cols and board[new_row][new_col] % 2:
                        live_cell += 1

                if (value % 2 and 2 <= live_cell <= 3) or (not value % 2 and live_cell == 3):
                    board[row][col] += 2
        
        for row in range(rows):
            for col in range(cols):
                board[row][col] //= 2
        